pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -278,17 +271,62 @@ func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
 	binary.BigEndian.PutUint32(icmpData[:ICMPUnusedHdrLen], 0)
 	copy(icmpData[ICMPUnusedHdrLen:], ipHdr[:ipHdrLen+8])
 	return c.ofClient.SendICMPPacketOut(
-		srcMAC.String(),
-		dstMAC.String(),
+		srcMAC,
+		dstMAC,
 		srcIP,
 		dstIP,
-		inPort,
-		-1,
+		uint32(inPort),
+		outPort,
 		isIPv6,
 		icmpType,
 		icmpCode,
 		icmpData,
-		true)
+		packetOutType)
+}
+
+// isServiceTraffic uses ServiceEPStateField to decided if this is a Service
+// traffic, when AntreaProxy is enabled. And when AntreaProxy is disabled, if the
+// IP of reject response packet is on local Node, but the MAC is MAC of antrea-gw,
+// then this is a rejection of Service traffic. Because this response is heading to
+// kube-proxy.
+func (c *Controller) isServiceTraffic(pktIn *ofctrl.PacketIn, antreaProxyEnabled bool, dstMAC string, isDstLocal bool) bool {
+	if antreaProxyEnabled {
+		matches := pktIn.GetMatches()
+		match := getMatchRegField(matches, openflow.ServiceEPStateField)
+		if match != nil {
+			svcEpstate, err := getInfoInReg(match, openflow.ServiceEPStateField.GetRange().ToNXRange())
+			return err == nil && svcEpstate&openflow.EpSelectedRegMark.GetValue() == openflow.EpSelectedRegMark.GetValue()
+		}
+	}
+	gwIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.GatewayInterface)
+	return isDstLocal && dstMAC == gwIfaces[0].MAC.String()
+}
+
+// getRejectOFPort sets the inPort and outPort of a packetOut based on the packetOutType.
+func getRejectOFPort(packetOutType binding.PacketOutType, sIface, dIface *interfacestore.InterfaceConfig) (int32, int32) {
+	inPort := int32(config.HostGatewayOFPort)
+	outPort := int32(-1)
+	switch packetOutType {
+	case binding.RejectPodLocal:
+		inPort = sIface.OFPort
+		outPort = dIface.OFPort
+	case binding.RejectServiceLocal:
+		inPort = sIface.OFPort
+	case binding.RejectPodRemoteToLocal:
+		inPort = config.HostGatewayOFPort
+		outPort = dIface.OFPort
+	case binding.RejectServiceRemoteToLocal:
+		inPort = config.HostGatewayOFPort
+	case binding.RejectLocalToRemote:
+		inPort = sIface.OFPort
+	case binding.RejectLocalToProxy:
+		inPort = sIface.OFPort
+		outPort = config.HostGatewayOFPort
+	case binding.RejectRemoteToProxy:
+		inPort = config.DefaultTunOFPort
```

Comment:
In noEncap mode, the inPort number should be "gatewayPort" because the TunOFPort doesn't exist?

Reply:
Sorry, my const name is bad. I already updated it and added comments. 
Here I use `DefaultTunOFPort` is because when AntreaProxy is disabled the encapMode must be encap.
There is other thing that I need to mention: For inter-node Pod-to-Pod traffic, when reject response is from remote to local, we couldn't only use encapMode to decide the inPort. Because in hybrid mode, we also need to check if the srcIP and dstIP are in the same subnet. So I only use gateway OFPort for rejection from remote to local Pod-to-Pod traffic. I wonder if this works for you guys. /cc @antoninbas .

Reply:
Actually, I don't get your point about `when AntreaProxy is disabled the encapMode must be encap`. So you mean if Antrea runs in noEncap/Hybrid mode, AntreaProxy must be enabled?

Reply:
Yes, our manifest generation has this restriction. https://github.com/antrea-io/antrea/blob/main/hack/generate-manifest.sh#L247-L249